baseapp/baseapp.go

```diff
@@ -761,3 +768,109 @@ func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*s
+		storeKeyName := storeKey.Name()
+		storeKeyNames = append(storeKeyNames, storeKeyName)
+		storeToLoadFrom[storeKeyName] = previousCMS.GetKVStore(storeKey)
+		storeKeyToSubstoreTraceBuf[storeKeyName] = &bytes.Buffer{}
+	}
+
+	// BaseApp, B1
+	options := []AppOption{
+		SetSubstoreTracer(storeTraceBuf),
+		SetTracerFor(storeKeys[0].Name(), storeKeyToSubstoreTraceBuf[storeKeys[0].Name()]),
+	}
+	newApp, err := SetupBaseAppFromParams(app.name, app.logger, dbm.NewMemDB(), app.txDecoder, storeKeyNames, app.router, app.LastBlockHeight()+1, storeToLoadFrom, options...)
+
+	return newApp, storeKeyToSubstoreTraceBuf, err
+}
+
+// Generate a fraudproof for an app with the given trace buffers
+func (app *BaseApp) generateFraudProof(storeKeyToSubstoreTraceBuf map[types.StoreKey]*bytes.Buffer) (FraudProof, error) {
+	fraudProof := FraudProof{}
+	fraudProof.stateWitness = make(map[string]StateWitness)
+	cms := app.cms.(*multi.Store)
+
+	for storeKey, subStoreTraceBuf := range storeKeyToSubstoreTraceBuf {
+		keys := cms.GetKVStore(storeKey).(*tracekv.Store).GetAllKeysUsedInTrace(*subStoreTraceBuf)
+
+		substoreSMT := cms.GetSubstoreSMT(storeKey.Name())
+		proof, storeHash, err := cms.GetSubStoreProof(storeKey.Name())
+		if err != nil {
+			return FraudProof{}, err
+		}
+		stateWitness := StateWitness{
+			proof:       *proof,
+			rootHash:    storeHash,
+			WitnessData: make([]WitnessData, 0, keys.Len()),
+		}
+		for key := range keys {
+			value := substoreSMT.Get([]byte(key))
+			proofOps, err := substoreSMT.GetProof([]byte(key))
+			if err != nil {
+				return FraudProof{}, err
+			}
+			proofOp := proofOps.GetOps()[0]
+			bKey, bVal := []byte(key), []byte(value)
+			witnessData := WitnessData{bKey, bVal, proofOp}
+			stateWitness.WitnessData = append(stateWitness.WitnessData, witnessData)
+		}
+		fraudProof.stateWitness[storeKey.Name()] = stateWitness
+	}
+
+	return fraudProof, nil
+}
+
+// set up a new baseapp from given params
+func SetupBaseAppFromParams(appName string, logger log.Logger, db dbm.Connection, txDecoder sdk.TxDecoder, storeKeyNames []string, router sdk.Router, blockHeight int64, storeToLoadFrom map[string]types.KVStore, options ...AppOption) (*BaseApp, error) {
+	storeKeys := make([]types.StoreKey, 0, len(storeKeyNames))
+	for _, storeKeyName := range storeKeyNames {
+		storeKey := sdk.NewKVStoreKey(storeKeyName)
+		storeKeys = append(storeKeys, storeKey)
+		subStore := storeToLoadFrom[storeKeyName]
+		it := subStore.Iterator(nil, nil)
+		for ; it.Valid(); it.Next() {
+			key, val := it.Key(), it.Value()
+			options = append(options, SetSubstoreKVPair(storeKey, key, val))
+		}
+	}
+	options = append(options, SetSubstores(storeKeys...))
+
+	// This initial height is used in `BeginBlock` in `validateHeight`
+	options = append(options, SetInitialHeight(blockHeight))
+
+	// make list of options to pass by parsing fraudproof
+	app := NewBaseApp(appName, logger, db, txDecoder, options...)
+
+	app.router = router
```

Comment:
what’s the best way to reuse a route from a one app B1 to another app B2? Just passing in the same router from B1 doesn’t seem to work  because during the route it uses B1 but I want it to use B2.

Reply:
not 100% sure of the context here, and just to clarify, a route or a router?

Reply:
when creating fraud proofs, we're creating an entirely new app, correct? so why do we need to pass the same router?

Reply:
Yes, we're creating an entirely new app but functionally we'd want the same routing method.

Reply:
sorry, I'm out of the loop, isn't the router stateless? what's the reason we need the same exact one?

also, in most cases we'll be working with the msgServiceRouter, correct?

Reply:
or I guess we'll need both if we want to support legacy stuff

Reply:
sorry I should clarify more, yes the router is stateless. But we'd still like the same function as routes to be used in the new baseapp. No state just logically. 

Yes in most cases we'll be using `msgServiceRouter`, is that also something I should initialize using the previous app's `msgServiceRouter`?